src/riak_kv_wm_timeseries.erl

```diff
@@ -29,572 +34,375 @@
+                                                                RD),
+                    {{halt, 400}, Resp, Ctx}
+            end
+    catch
+        throw:{data_problem,Reason} ->
+            Resp = riak_kv_wm_ts_util:set_error_message("wrong body: ~p", [Reason], RD),
+            {{halt, 400}, Resp, Ctx}
     end.
 
-
--spec extract_data([byte()]) -> undefined|any().
-extract_data(Json) ->
-    try mochijson2:decode(Json) of
-        Decoded when is_list(Decoded) ->
-            validate_ts_records(Decoded)
+-spec delete_resource(#wm_reqdata{}, #ctx{}) ->  cb_rv_spec(boolean()|halt()).
+delete_resource(RD,  #ctx{table=Table,
+                          mod=Mod,
+                          key=Key,
+                          options=Options}=Ctx) ->
+     try riak_kv_ts_api:delete_data(Key, Table, Mod, Options) of
+        ok ->
+             Json = result_to_json(ok),
+              Resp = riak_kv_wm_ts_util:set_json_response(Json, RD),
+              {true, Resp, Ctx};
+         {error, notfound} ->
+              {{halt, 404}, RD, Ctx}
+     catch
+         _:Reason ->
+             lager:log(info, self(), "delete_resource failed: ~p", [Reason]),
+             Resp = riak_kv_wm_ts_util:set_error_message("Internal error: ~p", [Reason], RD),
+             {{halt, 500}, Resp, Ctx}
+     end.
+
+-spec to_json(#wm_reqdata{}, #ctx{}) ->  cb_rv_spec(iolist()|halt()).
+to_json(RD, #ctx{api_call=get, object=Object}=Ctx) ->
+    try
+        Json = mochijson2:encode(Object),
+        {Json, RD, Ctx}
     catch
-        _:_ ->
-            undefined
+        _:Reason ->
+            Resp = riak_kv_wm_ts_util:set_error_message("object error ~p", [Reason], RD),
+            {{halt, 500}, Resp, Ctx}
     end.
 
-
-validate_ts_record(undefined) ->
-    undefined;
-validate_ts_record(R) when is_list(R) ->
-    case lists:all(
-           %% check that all list elements are TS types
-           fun(X) -> is_integer(X) orelse is_float(X) orelse is_binary(X) end,
-           R) of
-        true ->
-            R;
-        false ->
-            undefined
-    end;
-validate_ts_record(_) ->
-    undefined.
-
-validate_ts_records(RR) when is_list(RR) ->
-    case lists:all(fun(R) -> validate_ts_record(R) /= undefined end, RR) of
-        true ->
-            RR;
-        false ->
-            undefined
+-spec extract_params([{string(), string()}], #ctx{}) -> #ctx{} .
+%% @doc right now we only allow a timeout parameter or nothing.
+extract_params([], Ctx) ->
+    Ctx#ctx{options=[]};
+extract_params([{"timeout", TimeoutStr}], Ctx) ->
+    try
+        Timeout = list_to_integer(TimeoutStr),
+        Ctx#ctx{timeout = Timeout,
+                options = [{timeout, Timeout}]}
+    catch
+        _:_ ->
+            throw(riak_kv_wm_ts_util:flat_format("timeout not an integer value: ~s", [TimeoutStr]))
     end;
-validate_ts_records(_) ->
-    undefined.
+extract_params(Params, _Ctx) ->
+    throw(riak_kv_wm_ts_util:flat_format("incorrect paramters: ~p", [Params])).
```

Comment:
A utility function used several places should be in its own module. @jonmeredith - what are your thoughts?

I will try to avoid using it here as it might be good enough to use an io_lib:format/2.